crates/sui-light-client/src/main.rs

```diff
@@ -0,0 +1,680 @@
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+async fn get_full_checkpoint(config: &Config, seq: u64) -> anyhow::Result<CheckpointData> {
+    // Downloading the checkpoint from the server
+    let client: Client = Client::new(config.rest_url());
+    let full_checkpoint = client.get_full_checkpoint(seq).await?;
+
+    Ok(full_checkpoint)
+}
+
+fn extract_verified_effects_and_events(
+    checkpoint: &CheckpointData,
+    committee: &Committee,
+    tid: TransactionDigest,
+) -> anyhow::Result<(TransactionEffects, Option<TransactionEvents>)> {
+    let summary = &checkpoint.checkpoint_summary;
+
+    // Verify the checkpoint summary using the committee
+    summary.verify_with_contents(committee, Some(&checkpoint.checkpoint_contents))?;
+
+    // Check the validity of the transaction
+    let contents = &checkpoint.checkpoint_contents;
+    let (matching_tx, _) = checkpoint
+        .transactions
+        .iter()
+        .zip(contents.iter())
+        // Note that we get the digest of the effects to ensure this is
+        // indeed the correct effects that are authenticated in the contents.
+        .find(|(tx, digest)| {
+            tx.effects.execution_digests() == **digest && digest.transaction == tid
+        })
+        .ok_or(anyhow!("Transaction not found in checkpoint contents"))?;
+
+    // Check the events are all correct.
+    let events_digest = matching_tx.events.as_ref().map(|events| events.digest());
+    anyhow::ensure!(
+        events_digest.as_ref() == matching_tx.effects.events_digest(),
+        "Events digest does not match"
+    );
+
+    // Since we do not check objects we do not return them
+    Ok((matching_tx.effects.clone(), matching_tx.events.clone()))
+}
+
+async fn get_verified_effects_and_events(
+    config: &Config,
+    tid: TransactionDigest,
+) -> anyhow::Result<(TransactionEffects, Option<TransactionEvents>)> {
+    let sui_mainnet: Arc<sui_sdk::SuiClient> = Arc::new(
+        SuiClientBuilder::default()
+            .build(config.full_node_url.as_str())
+            .await
+            .unwrap(),
+    );
+    let read_api = sui_mainnet.read_api();
+
+    // Lookup the transaction id and get the checkpoint sequence number
+    let options = SuiTransactionBlockResponseOptions::new();
+    let seq = read_api
+        .get_transaction_with_options(tid, options)
+        .await?
+        .checkpoint
+        .ok_or(anyhow!("Transaction not found"))?;
+
+    // Download the full checkpoint for this sequence number
+    let full_check_point = get_full_checkpoint(config, seq).await?;
+
+    // Load the list of stored checkpoints
+    let checkpoints_list: CheckpointsList = read_checkpoint_list(config)?;
+    // find the stored checkpoint before the seq checkpoint
+    let prev_ckp_id = checkpoints_list
+        .checkpoints
+        .iter()
+        .filter(|ckp_id| **ckp_id < seq)
+        .last()
+        .ok_or(anyhow!("No checkpoint found before the transaction"))?;
```

Comment:
if no checkpoint is found, means its immediately after genesis?

Reply:
This is the idea.

Reply:
Actually looking at this, if we do not find a checkpoint this is meant to return `Err`. There might be a boundary issue here.

Reply:
Great find, this was incorrectly dealing with checkpoints signed with the genesis committee. Now fixed to use genesis if no checkpoint was found in the list.